### PR TITLE
Fixed crash of pagination

### PR DESCRIPTION
**Pull request by gipfelen**

### CODE REVIEW

**Comment by gipfelen** [comment]
That’s exactly true. I encountered this bug while experimenting with the vuetify data table, which returns -1 for the itemsPerPage parameter if we select “all”.

But even though if we wouldn’t support to get all elements, a possible attacker could achieve the same thing by passing e.g. 1 million for this parameter. Thus, we would need to add an upper limit globally.

**Comment by tanner0101** [comment]
Yes that's a good point. This needs more thought. To fix the immediate problem, I'd prefer to take the simplest possible approach here. As I mentioned:

> The easiest way to do that (without adding new functionality) would be to check if request.start < request.end or throw an error.

Then a subsequent PR can be made to loosen this restriction and offer some configurability around maximum page lengths and whether or not `-1` should be allowed to specify "all". I'd rather do it this way as opposed to making `-1` work now by default then potentially taking this feature away in a future patch.

**Comment by gipfelen** [comment]
I've changed the PR to only solve the immediate problem.

File: Sources/FluentKit/FluentError.swift [hunk 1]
```diff
@@ -22,6 +23,8 @@ public enum FluentError: Error, LocalizedError, CustomStringConvertible {
             return "invalid field: \(name) type: \(valueType) error: \(error)"
         case .noResults:
             return "Query returned no results"
+        case .invalidPagination:
+            return "Invalid parameters for pagination"
```
Comment (texuf) [inline Sources/FluentKit/FluentError.swift:27]: you added this new error, buy you're not using it... is that intentional?

**Comment by gipfelen** [comment]
ooo I forgot to use it... but now i do

**Comment by texuf** [comment]
If pagination "just doesn't work" if we pass 0 for page or for per, then this seems like a good addition to me.

**Comment by 0xTim** [comment]
Can we have a test to prove it no longer crashes?

**Comment by gipfelen** [comment]
Here you go

**Comment by 0xTim** [comment]
Making the initialiser throwing is a breaking change that would require a major release. I'm not sure the best way around it. We could deprecate the existing initialiser and bound the values so it can't crash and use the new throwing initialiser as the deprecation fix

**Comment by gipfelen** [comment]
So we have to change the signature for the new throwing initializer, right? Any suggestions what it should look like?

**Review by siemensikkema — COMMENTED** [review]
@gipfelen It seems to me there is no good way to have both a throwing and a non-throwing, albeit deprecated, initializer. 
@tanner0101 's suggestions included checking for `request.start < request.end` which would be in `QueryBuilder+Paginate.swift`. There we could just return an empty page if start > end.

Alternatively, instead of throwing, we could just do the bounding that @0xTim suggested? So `self.page = max(1, page)` and `self.per = max(1, per)` for example but that would have to be done in 2 places and then the metadata returned does not reflect the actual request.

**Comment by gwynne** [comment]
Closing due to lack of activity. Please feel free to open a new PR if this is still something you want to see addressed!

**Pull request closed without merging**